plugins/huggingface/modelgauge/suts/huggingface_client.py

```diff
@@ -290,130 +199,78 @@ def evaluate(self, raw_request: HuggingFaceRequest) -> HuggingFaceResponse:
-                with self.wrapped_tokenizer as tokenizer:
-                    generated_tokens_top_logprobs_dicts.append(
-                        {
-                            tokenizer.convert_ids_to_tokens(k.item()): v.item()
-                            for (k, v) in zip(
-                                topk_logprobs.indices, topk_logprobs.values
-                            )
-                        }
-                    )
-                # Get log probability of chosen token.
-                j = i + len(encoded_input.input_ids[0])
-                generated_tokens_logprobs.append(
-                    logprobs[sequences[completion_id][j]].item()
-                )
-            all_generated_tokens_logprobs.append(generated_tokens_logprobs)
-            all_generated_tokens_top_logprobs_dicts.append(
-                generated_tokens_top_logprobs_dicts
-            )
-
-        # Remove prompt from the start of each sequence.
-        sequences = [
-            sequence[len(encoded_input.input_ids[0]) :] for sequence in sequences
-        ]
-
-        with self.wrapped_tokenizer as tokenizer:
-            all_tokens = [
-                [tokenizer.decode(token) for token in sequence_tokens]
-                for sequence_tokens in sequences
-            ]
-            all_decoded_text = tokenizer.batch_decode(sequences)
+                optional_args["stopping_criteria"] = stopping_criteria
+        # Some models do not have a `pad_token_id`. For example gpt2
+        # This prevents a warning message
+        if self.model.config.pad_token_id is None:
+            if self.model.config.eos_token_id is not None:
+                optional_args["pad_token_id"] = self.model.config.eos_token_id
+        output_scores = (
+            raw_request.num_top_logprobs is not None
+            and raw_request.num_top_logprobs > 0
+        )
+        output = self.model.generate(
+            **encoded_input,
+            **raw_request.generate_args.model_dump(),
+            max_new_tokens=max_new_tokens,
+            do_sample=True,
+            return_dict_in_generate=True,
+            output_scores=output_scores,
+            **optional_args,
+        )
 
         completions: List[HuggingFaceCompletion] = []
-        for (
-            decoded_text,
-            tokens,
-            generated_tokens_logprobs,
-            generated_tokens_top_logprobs_dicts,
-        ) in zip(
-            all_decoded_text,
-            all_tokens,
-            all_generated_tokens_logprobs,
-            all_generated_tokens_top_logprobs_dicts,
-        ):
+        for completion_id in range(raw_request.generate_args.num_return_sequences):
+            # Remove input prompt and truncate if length exceeds max_new_tokens.
+            sequence = output.sequences[completion_id][num_input_tokens:]
+            sequence = sequence[:max_new_tokens]
+
+            text = tokenizer.decode(sequence)
+            sequence_logprobs_dicts = None
```

Comment:
Is it possible to just initialize this with an empty list instead of having it be `None` here and an empty (or not) list later?

Reply:
So the idea behind the ProducesPerTokenLogProbs capability is that logprobs should only be returned if they are specifically requested. And I think `None` makes it explicit that the completion does not include logprobs, whereas an empty list is more ambiguous (e.g. Were logprobs not requested or was there an error in getting logprobs that caused to be empty?)